### DISCUSSION TITLE
Notes on the IJsselbridge as a use-case for probeye

### DISCUSSION THREAD

**Discussion by JanKoune** [BAMresearch/probeye#49] [General]
This issue contains some notes on the IJsselbridge use-case, where taralli is used to perform Bayesian system identification. The aim is to identify parts of the workflow where the current implementation (using taralli or otherwise) can be replaced with probeye. For ease of reading this issue is separated into the following sections:

1. Problem description
2. Sensors
3. Parameters and priors
4. Likelihood
4. Additional notes

## 1. Problem description

An overview of the IJsselbridge use-case is provided in the figure below:

<img src="https://user-images.githubusercontent.com/54745276/133930116-6d19fe90-f458-452b-8ea5-f87f4ec803ad.PNG" width="600" />

The IJsselbridge is a twin girder steel road bridge, modeled by a Euler-Bernoulli beam FE model. This physical model is combined with a probabilistic model describing the modeling and measurement uncertainties. The measurement uncertainty at the measurement locations are taken as i.i.d. Gaussian random variables with zero mean. The modeling uncertainty is modeled as a multivariate Gaussian distribution, with the uncertainty at each point scaled by the model output, and correlation defined by a chosen kernel function (e.g. Exponential). The aim of the IJsselbridge use-case is to perform:

* Parameter estimation for uncertain parameters of the FE model, e.g. the rotational stiffness of the supports and the stiffness of the K-braces connecting the two girders.
* Parameter estimation for the uncertain parameters of the probabilistic model that defines the modeling and measurement uncertainties.
* Model selection for a pool of candidate coupled physical-probabilistic models, each with a different correlation function (and possibly physical model).

## 2. Sensors

The data is obtained from strain gauges located at the bottom flange of a girder that measure the change in stress at positions along the bridge as a truck with known weight drives across the bridge at a constant speed. Each sensor is defined by a single spatial coordinate `x` : it's position along the bridge length. Every measurement can then be indexed by the coordinate `x` of the sensor and the coordinate `t` of the measurement:

<img src="https://user-images.githubusercontent.com/54745276/133932180-b37acb47-70eb-4dc3-a929-fde656210200.png" width="300" />

Definition of the sensors and the additive measurement uncertainty component could be handled by the existing probeye functionality.

## 3. Parameters and priors

Currently the priors are defined as shown below:

```Python

# A solver-specific example of the priors used in the IJsselbridge case:
# Dynesty requires the prior to be specified in terms of the inverse CDF, taking samples
# from the unit cube and transforming them into samples from the prior distribution.

# Example prior
def prior(u):
    x = np.empty(n_theta)
    x[0] = stats.norm.ppf(u[0], loc = a11, scale = b11)
    x[1] = stats.norm.ppf(u[1], loc = a12, scale = b12)
    x[2] = stats.norm.ppf(u[2], loc = a13, scale = b13)
    # ...
    return x
```

This is prone to error when defining several models with a relatively large number of probabilistic and physical parameters (e.g. > ~10). Probeye could be used to make the definition of parameters and priors more intuitive and less error-prone by:

* Defining priors on the level of each parameter.
* Distinguishing between physical model parameters and uncertainty parameters.
* Handling parameters by name rather than index.

## 4. Likelihood
The vector of unknown parameters is composed of the vector of physical model parameters (denoted by the subscript **_s_**) and probabilistic model parameters (subscript **_c_**). The data generating process described in the introduction and the resulting likelihood function are given below.

<img src="https://user-images.githubusercontent.com/54745276/134906745-af25fc1b-1f27-4994-9bda-5580296328bd.PNG" width="600" />

The physical model parameterization can vary, but a representative example would be 4 rotational stiffness parameters, _θc_ = {Kr1, Kr2, Kr3, Kr4} with uniform priors in the range [0, 10^8] kNm/rad, representing the stiffness of the first four supports of the bridge:

<img src="https://user-images.githubusercontent.com/54745276/134901362-72dce619-3f33-4759-96ec-4d11efed7125.PNG" width="500" />

The probabilistic model parameters include the modeling uncertainty coefficient of variation _Cv_ *, the parameters of the specified kernel (e.g. correlation lengthscale), and possibly the measurement uncertainty standard deviation _σmeas_. Currently, uniform priors are specified for these parameters.

*Note: The modeling uncertainty is included in the kernel function in place of the marginal variance parameter _σ_ (e.g. as in https://www.cs.toronto.edu/~duvenaud/cookbook/). It is scaled by the model prediction at each point, meaning that it can be taken as a coefficient of variation.

## 5. Additional notes

The following is a list of potential limitations on using probeye for the IJsselbridge:

1. Correlation in the model prediction error is assumed.
2. Multiplicative modeling uncertainty and additive measurement uncertainty is assumed.
3. Inference is performed using nested sampling.

**Comment by joergfunger** [comment]
@JanKoune could you add the likelihood you want to use in this example, including the (hyper-)parameters to be estimated (assuming not all parameters are normal distributed?

**Comment by JanKoune** [comment]
@joergfunger I have included a section detailing the likelihood function and parameters. If you need any additional information, please let me know.

**Comment by atulag0711** [comment]
@JanKoune Thanks for the interesting use case. I have a few questions (some maybe very basic):
1. Now you have used uniform priors for the \theta_c. Can there be some prior knowledge on spatial/temporal correlation and if it exists will it be problem specific for eg. smoothness priors in space, correlation as a function of temporal step size (Just a guess) etc?
2. Why choose multiplicative noise for model uncertainty? Why not additive for example. Intuitive explanation would help. 
3. I think propbeye follows hierarchical bayesian structure, So prior for cov kernel can be included and the noise type changed from normal to MVN with a provision to include the cov kernel.

**Comment by JanKoune** [comment]
Hi @atulag0711,

Some notes on your questions, to the best of my understanding:

1. I think that prior information regarding the uncertainty, including the type of kernel that can be used to model the correlation and the prior distribution of the kernel parameters, will likely be problem-specific. In this case engineering judgement is applied to determine the bounds of the parameters, e.g. the marginal variance and correlation lengthscale, by specifying uniform priors. This seems to be the typical approach in defining priors for such parameters, using e.g. uniform or inverse-gamma distributions. There are some publications that discuss a more rigorous definition of priors for kernel parameters (e.g. [Fuglstad et. al](https://www.researchgate.net/publication/273067384_Interpretable_Priors_for_Hyperparameters_for_Gaussian_Random_Fields))  but it is not a subject I am very familiar with.

2. For this particular problem we would expect the modeling uncertainty to be dependent on the magnitude of the model output. Considering the stress influence line obtained for a location in the FE model shown previously: For a unit load on top of (or near) a support, the stress will be (close to) zero as specified by the boundary conditions. The modeling uncertainty in this case is smaller compared to the case of a load acting on the mid-span, whereas under the assumption of additive uncertainty, both these model predictions would have the same uncertainty. The multiplicative uncertainty also accounts for effects that would scale the influence line, e.g. offset of the load transverse position from the one assumed in the model.

3. Thank you for the information! My impression from briefly looking through the code was that this may be possible by extending the `NoiseModelBase` class, but it may be that other changes are needed as well.
